crates/ra_assists/src/handlers/change_return_type_to_result.rs

```diff
@@ -0,0 +1,798 @@
+            }
+        },
+    )
+}
+
+struct TailReturnCollector {
+    exprs_to_wrap: Vec<SyntaxNode>,
+}
+
+impl TailReturnCollector {
+    fn new() -> Self {
+        Self { exprs_to_wrap: vec![] }
+    }
+    /// Collect all `break` and `return` expression
+    fn collect_return_exprs(&mut self, block_expr: &Block) {
+        let statements = block_expr.statements();
+        for stmt in statements {
+            let expr_stmt = match &stmt {
+                ast::Stmt::ExprStmt(stmt) => stmt.expr(),
+                ast::Stmt::LetStmt(stmt) => stmt.initializer(),
+            };
+            if let Some(expr) = &expr_stmt {
+                self.handle_exprs(expr);
+            }
+        }
+
+        // Browse tail expressions for each block
+        if let Some(expr) = block_expr.expr() {
+            if let Some(last_exprs) = get_tail_expr_from_block(&expr) {
+                for last_expr in last_exprs {
+                    if let Some(block) = Block::cast(last_expr.clone()) {
+                        self.collect_return_exprs(&block);
+                    } else if let Some(last_expr) = Expr::cast(last_expr.clone()) {
+                        self.handle_exprs(&last_expr);
+                    } else if let Some(expr_stmt) = ast::Stmt::cast(last_expr) {
+                        let expr_stmt = match &expr_stmt {
+                            ast::Stmt::ExprStmt(stmt) => stmt.expr(),
+                            ast::Stmt::LetStmt(stmt) => stmt.initializer(),
+                        };
+                        if let Some(expr) = &expr_stmt {
+                            self.handle_exprs(expr);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn handle_exprs(&mut self, expr: &Expr) {
+        match expr {
+            Expr::BlockExpr(block_expr) => {
+                if let Some(block) = block_expr.block() {
+                    self.collect_return_exprs(&block);
+                }
+            }
+            Expr::ReturnExpr(ret_expr) => {
+                if let Some(ret_expr_arg) = &ret_expr.expr() {
+                    self.exprs_to_wrap.push(ret_expr_arg.syntax().clone());
+                }
+            }
+            Expr::BreakExpr(break_expr) => {
+                if let Some(break_expr_arg) = &break_expr.expr() {
+                    self.exprs_to_wrap.push(break_expr_arg.syntax().clone());
+                }
+            }
+            Expr::IfExpr(if_expr) => {
+                for block in if_expr.blocks() {
+                    if let Some(in_block) = block.block() {
+                        self.collect_return_exprs(&in_block);
+                    }
+                }
+            }
+            Expr::LoopExpr(loop_expr) => {
+                if let Some(last_child) = loop_expr.syntax().last_child() {
+                    if let Some(block_expr) = BlockExpr::cast(last_child) {
```

Comment:
You should be able to use `loop_body` for this. Same for `ForExpr` and `WhileExpr`.